### PR TITLE
Uncertainty review

### PR DESCRIPTION
**Pull request by ThierryJudge**
Add a review of uncertainty estimation methods.

### CODE REVIEW

File: blog/_posts/2021-06-17-uncertainty.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+$$
+\mathbb{P} ( \hat{Y} = Y | \hat{P} = P ) = p  \text{   } \forall p \in [0, 1]
+$$
+
+where $$\hat{Y}$$ is the predicted class and $$\hat{P}$$ is the predicted probability.
+
+Calibration is often shown with reliability diagrams [^5] (see figures in the following sections). These diagrams are built by 
+splitting the samples into bins according to their confidence. The accuracy for each bin each calculated and should 
+correspond to the average confidence. The difference between the accuracy and confidence is known as the gab. Averaging
```
Comment (jhlegarreta) [inline blog/_posts/2021-06-17-uncertainty.md:28]: gab : gap

There are multiple instances of this typo.

File: blog/_posts/2021-06-17-uncertainty.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+the gab for all bins gives the expected calibration error (ECE) [^6]. 
+
+## Uncertainty-error overlap
+In the case of segmentation, the uncertainty of a prediction can be predicted via an uncertainty map. An uncertainty map is an
+uncertainty prediction for each pixel. Uncertainty map quality can be evaluated by computing the uncertainty-error 
+overlap. Uncertainty-error overlap is computed by obtaining the dice score of the uncertainty map and a pixel-wise error
+map between the prediction and the groundtruth segmentation map.
```
Comment (jhlegarreta) [inline blog/_posts/2021-06-17-uncertainty.md:35]: A picture would help here.

File: blog/_posts/2021-06-17-uncertainty.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+# Evaluating Uncertainty
+
+## Model calibration
+
+A neural network trained for classification is said to be calibrated if its confidence is equal to the probability of 
+it being correct. For example, given 100 predictions with an average confidence of 0.7, 70 predictions should be correct. 
+This can be expressed :
+
+$$
+\mathbb{P} ( \hat{Y} = Y | \hat{P} = P ) = p  \text{   } \forall p \in [0, 1]
```
Comment (jhlegarreta) [inline blog/_posts/2021-06-17-uncertainty.md:21]: Was the `\text{   }` statement intended to contain something? If you want to introduce whitespaces in math mode, there are some specific commands to do so `\;`, `\quad`: https://www.overleaf.com/learn/latex/Spacing_in_math_mode

File: blog/_posts/2021-06-17-uncertainty.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+* The entropy of the vector: $$\sum_{i=0}^C \vec{p}_i \log(\vec{p}_i)$$. The higher the entropy, the higher the 
+uncertainty as the entropy will be at its highest if all classes are equally probable $$ \vec{p}_i = \frac{1}{C}$$.
+
+In the case of a simple classification output, this results in one uncertainty value per classification. In the case of 
+segmentation, the uncertainty evaluation is done for each pixel and results in an uncertainty map with one real value per
+pixel.  
+
+### Temperature Scaling
+
+Guo et al. show that neural networks, especially deep networks, tend to be miscalibrated as can be seen in the following
```
Comment (jhlegarreta) [inline blog/_posts/2021-06-17-uncertainty.md:56]: Missing the cross-ref for Guo et al.

File: blog/_posts/2021-06-17-uncertainty.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+#### What uncertainties are needed?
+Kendall et al. describe two types of uncertainty in *What Uncertainties Do We Need in Bayesian Deep
+Learning for Computer Vision?* [^7]. Firstly, epistemic uncertainty is the uncertainty in the model weights. This can be
+represented with Bayesian nets or MC dropout. This type of uncertainty can be decreased with more data. The second type 
+of uncertainty is aleatoric uncertainty; the uncertainty in the data. This type of uncertainty cannot be 
+reduced by more data as it is inherent to the data. Aleatoric uncertainty can be homoscedastic (constant for all 
+inputs) or heteroscedastic (dependant on the input sample). 
+
+As mentioned estimating epistemic uncertainty can be done using MC dropout. The prediction is given by:
```
Comment (jhlegarreta) [inline blog/_posts/2021-06-17-uncertainty.md:141]: `As mentioned estimating epistemic uncertainty can be done (...)`: `As mentioned, decreasing epistemic uncertainty can be done (...)` ?

File: blog/_posts/2021-06-17-uncertainty.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+$$
+L(\theta) = \frac{1}{N} \sum_{i=1}^N \frac{1}{2\sigma^w(x_i)^2} ||y_i - f^w(x_i)||^2 + \frac{1}{2}\sigma^w(x_i)^2
+$$
+
+
+![](/blog/images/uncertainty/aleatoric.jpg)[^3]
+
+
+In the case of classification, the formulation is slightly more complicated. Given a model logits $$f^w(x_i)$$ and 
+variance prediction $$\sigma^w(x_i)$$, the logits can be corrupted by Gaussian nose with variance $$\sigma^w(x_i)^2$$.
```
Comment (jhlegarreta) [inline blog/_posts/2021-06-17-uncertainty.md:171]: Gaussian nose : Gaussian noise

File: blog/_posts/2021-06-17-uncertainty.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+[^7]: Kendall, A., & Gal, Y. (2017). What Uncertainties Do We Need in Bayesian Deep Learning for Computer Vision? In I. Guyon, U. V. Luxburg, S. Bengio, H. Wallach, R. Fergus, S. Vishwanathan, & R. Garnett (Eds.), Advances in Neural Information Processing Systems (Vol. 30). Curran Associates, Inc. https://proceedings.neurips.cc/paper/2017/file/2650d6089a6d640c5e85b2b88265dc2b-Paper.pdf
+
+[^8]: Balaji Lakshminarayanan, Alexander Pritzel, and Charles Blundell. 2017. Simple and scalable predictive uncertainty estimation using deep ensembles. In Proceedings of the 31st International Conference on Neural Information Processing Systems (NIPS'17). Curran Associates Inc., Red Hook, NY, USA, 6405–6416.
+
+[^9]: Corbière, C., THOME, N., Bar-Hen, A., Cord, M., & Pérez, P. (2019). Addressing Failure Prediction by Learning Model Confidence. In H. Wallach, H. Larochelle, A. Beygelzimer, F. d\textquotesingleAlché-Buc, E. Fox, & R. Garnett (Eds.), Advances in Neural Information Processing Systems (Vol. 32). Curran Associates, Inc. https://proceedings.neurips.cc/paper/2019/file/757f843a169cc678064d9530d12a1881-Paper.pdf
```
Comment (jhlegarreta) [inline blog/_posts/2021-06-17-uncertainty.md:243]: \textquotesingle: probably slipped from the bibtex file.

**Review by jhlegarreta — COMMENTED** [review]
A few comments inline.

File: blog/_posts/2021-06-17-uncertainty.md (hunk 5 above)
Comment (ThierryJudge) [reply blog/_posts/2021-06-17-uncertainty.md:141]: estimating epistemic uncertainty is done with MC dropout, to reduce it you must add more data.

Comment (jhlegarreta) [reply blog/_posts/2021-06-17-uncertainty.md:144]: OK.

File: blog/_posts/2021-06-17-uncertainty.md (hunk 1 above)
Comment (jhlegarreta) [reply blog/_posts/2021-06-17-uncertainty.md]: Done in eacdc9f.

File: blog/_posts/2021-06-17-uncertainty.md (hunk 7 above)
Comment (ThierryJudge) [reply blog/_posts/2021-06-17-uncertainty.md]: Fixed in eacdc9f

File: blog/_posts/2021-06-17-uncertainty.md (hunk 3 above)
Comment (jhlegarreta) [reply blog/_posts/2021-06-17-uncertainty.md]: Done in eacdc9f.

File: blog/_posts/2021-06-17-uncertainty.md (hunk 6 above)
Comment (ThierryJudge) [reply blog/_posts/2021-06-17-uncertainty.md]: Fixed in eacdc9f

File: blog/_posts/2021-06-17-uncertainty.md (hunk 4 above)
Comment (ThierryJudge) [reply blog/_posts/2021-06-17-uncertainty.md]: Fixed in eacdc9f

File: blog/_posts/2021-06-17-uncertainty.md (hunk 6 above)
Comment (jhlegarreta) [reply blog/_posts/2021-06-17-uncertainty.md]: Done in eacdc9f.

File: blog/_posts/2021-06-17-uncertainty.md (hunk 2 above)
Comment (ThierryJudge) [reply blog/_posts/2021-06-17-uncertainty.md]: Is the picture added in eacdc9f clear ?

File: blog/_posts/2021-06-17-uncertainty.md (hunk 7 above)
Comment (jhlegarreta) [reply blog/_posts/2021-06-17-uncertainty.md]: Done in eacdc9f.

File: blog/_posts/2021-06-17-uncertainty.md (hunk 4 above)
Comment (jhlegarreta) [reply blog/_posts/2021-06-17-uncertainty.md]: Done in eacdc9f.

File: blog/_posts/2021-06-17-uncertainty.md (hunk 2 above)
Comment (jhlegarreta) [reply blog/_posts/2021-06-17-uncertainty.md]: Not sure if the thresholded uncertainty added in eacdc9f shows the uncertainty-error overlap, but anyways the description is clear having re-read it.

**Review by jhlegarreta — APPROVED** [review]

**Review by jodoin — APPROVED** [review]
